File: src/leetcode/linkedlist/ReverseList.java
package leetcode.linkedlist;

/**
 * @author chengzw
 * @description 反转链表，https://leetcode-cn.com/problems/reverse-linked-list/
 * @since 2021/8/27
 */
public class ReverseList {

    /**
     * 思路：反转操作
     * p.next = newNode
     * newNode = p
     * @param head
     * @return
     */
    public static ListNode solution1(ListNode head){
        ListNode p = head;
        //虚拟头节点
        ListNode newNode = null;

        while(p != null){
            ListNode tmp = p.next;
            p.next = newNode;
            newNode = p;
            p = tmp;
        }

        return newNode;
    }
}

File: go/leetcode/linkedlist/reverse_list.go
package linkedlist

/**
 * @description 206. 反转链表
 * @author chengzw
 * @since 2022/5/20
 * @link https://leetcode.cn/problems/reverse-linked-list/
 */

/**
给你单链表的头节点 head ，请你反转链表，并返回反转后的链表。
示例 1：
输入：head = [1,2,3,4,5]
输出：[5,4,3,2,1]
*/

/**
思路：反转操作
p.next = newNode
newNode = p

时间复杂度: O(n)
空间复杂度: O(1)
*/
func reverseList(head *ListNode) *ListNode {
	var newHead *ListNode
	p := head
	for p != nil {
		tmp := p.Next
		p.Next = newHead
		newHead = p
		p = tmp
	}
	return newHead
}
